hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/HoodieDeltaStreamerMultiwriterCheckpoint.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.exception.HoodieIOException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static org.apache.hudi.utilities.deltastreamer.HoodieDeltaStreamer.CHECKPOINT_KEY;
+
+/**
+ * This is used as an extraPreCommitFunc in BaseHoodieWriteClient
+ * It adds the checkpoint to deltacommit metadata. It must be implemented this way
+ * because it needs the lock to ensure that it does not overwrite another deltastreamers
+ * latest checkpoint with an older one.
+ */
+public class HoodieDeltaStreamerMultiwriterCheckpoint implements BiConsumer<HoodieTableMetaClient, HoodieCommitMetadata> {
+  private static final ObjectMapper OM = new ObjectMapper();
+
+  //deltastreamer id
+  private final String id;
+
+  //the deltastreamer
+  private final DeltaSync ds;
+  private final String checkpoint;
+  private final String latestCheckpointWritten;
+  
+  public HoodieDeltaStreamerMultiwriterCheckpoint(DeltaSync ds, String checkpoint, String latestCheckpointWritten) {
+    this.ds = ds;
+    this.id = ds.getId();
+    this.checkpoint = checkpoint;
+    this.latestCheckpointWritten = latestCheckpointWritten;
+  }
+
+  @Override
+  public void accept(HoodieTableMetaClient metaClient, HoodieCommitMetadata commitMetadata) {
+    //Get last completed deltacommit
+    Option<HoodieCommitMetadata> latestCommitMetadata;
+    try {
+      ds.refreshTimeline();
+      latestCommitMetadata = ds.getLatestCommitMetadataWithValidCheckpointInfo(metaClient.getActiveTimeline().getCommitsTimeline());
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to get the latest commit metadata", e);
+    }
+
+    //Get checkpoint map
+    Map<String,String> checkpointMap;
+    if (latestCommitMetadata.isPresent()) {
+      String value = commitMetadata.getMetadata(CHECKPOINT_KEY);
+      try {
+        checkpointMap = OM.readValue(value, Map.class);
+      } catch (IOException e) {
+        throw new HoodieIOException("Failed to parse checkpoint as map", e);
+      }
+    } else {
+      checkpointMap = new HashMap<>();
+    }
+
+    if (!StringUtils.isNullOrEmpty(latestCheckpointWritten)
+        && checkpointMap.containsKey(id) && !checkpointMap.get(id).equals(latestCheckpointWritten)) {
+      throw new HoodieException(String.format("Multiple DeltaStreamer instances with id: %s detected. Each Deltastreamer must have a unique id.", id));
+    }
+
+    //Add map to metadata
+    checkpointMap.put(id, checkpoint);
+    try {
+      commitMetadata.addMetadata(CHECKPOINT_KEY, OM.writeValueAsString(checkpointMap));
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public static Option<String> readCheckpointValue(String value, String id) {
```

Comment:
this is preCommitFunc. lets not place static method here. 
may be we can create DeltastreamerCheckpointUtils and move this method and any other static method related to checkpoint parsing.

Reply:
I just thought it would make more sense to keep this logic here. I moved it back into deltasync. Keeping it here also reduces the number of objectmappers used